nest_elephant_tvb/translation/mpi_translator.py

```diff
@@ -3,192 +3,237 @@
-    :param buffer: the buffer which contains the data (SHARED between thread)
-    :return:
+    Receive data on rank 0. Put it into the shared mem buffer.
+    Replaces the former 'receive' function.
+    NOTE: First refactored version -> not pretty, not final. 
     '''
-    # initialise variables for the loop
-    status_ = MPI.Status() # status of the different message
-    source_sending = np.arange(0,comm.Get_remote_size(),1) # list of all the process for the commmunication
+    status_ = MPI.Status()
+    num_sending = comm_receiver.Get_remote_size() # how many NEST ranks are sending?
+    # TODO: It seems the 'check' variable is used to receive tags from NEST, i.e. ready for send...
+    # change this in the future, also mentioned in the FatEndPoint solution from Wouter.
     check = np.empty(1,dtype='b')
-    count=0
-    while True: # FAT END POINT
-        # send the confirmation of the process can send data
+    shape = np.empty(1, dtype='i')    
+    count = 0
+    # TODO: the last two buffer entries are used for shared information
+    # --> they replace the status_data variable from previous version
+    # --> find more elegant solution?
+    databuffer[-1] = 1 # set buffer to 'ready to receive from nest'
+    databuffer[-2] = 0 # marks the 'head' of the buffer
+    
+    while(True):
         logger.info(" Nest to TVB : wait all")
-        for source in source_sending:
-            comm.Recv([check, 1, MPI.CXX_BOOL], source=source, tag=MPI.ANY_TAG, status=status_)
-
+        head_ = 0 # head of the buffer, reset after each iteration
+         
+        # TODO: This is still not correct. We only check for the Tag of the last rank.
+        # TODO: IF all ranks send always the same tag in one iteration (simulation step)
+        # TODO: then this works. But it should be handled differently!!!!
+        for i in range(num_sending):
+            # new: We do not care which source sends first, give MPI the freedom to send in whichever order.
+            comm_receiver.Recv([check, 1, MPI.CXX_BOOL], source=MPI.ANY_SOURCE, tag=MPI.ANY_TAG, status=status_)
+        # TODO: handle properly, all ranks send tag 0?
         if status_.Get_tag() == 0:
-            logger.info(" Nest to TVB : start to receive")
-            #  Get the data/ spike
-            for source in source_sending:
-                comm.Send([np.array(True,dtype='b'),MPI.BOOL],dest=source,tag=0)
-                shape = np.empty(1, dtype='i')
-                comm.Recv([shape, 1, MPI.INT], source=source, tag=0, status=status_)
-                data = np.empty(shape[0], dtype='d')
-                comm.Recv([data, shape[0], MPI.DOUBLE], source=source, tag=0, status=status_)
-                store.add_spikes(count,data)
-            while status_data[0] != 1 and status_data[0] != 2: # FAT END POINT
+            # wait until ready to receive new data (i.e. the sender has cleared the buffer)
+            while databuffer[-1] != 1:
                 time.sleep(0.001)
```

Comment:
With MPI would make this a hot loop: 
I would add an TODO: this is a potential place for cheap performance gains (remove the sleep)

Reply:
Marked as TODO in both loops, receiving and sending.